File: computing/interview/data-structures/cpp/stack.cpp
// Stack — using std::stack (adaptor over std::deque by default)
#include <iostream>
#include <stack>
#include <string>

// Classic use-case: balanced parentheses checker
bool isBalanced(const std::string& s) {
    std::stack<char> st;
    for (char c : s) {
        if (c == '(' || c == '[' || c == '{') {
            st.push(c);
        } else if (c == ')' || c == ']' || c == '}') {
            if (st.empty()) return false;
            char top = st.top(); st.pop();
            if ((c == ')' && top != '(') ||
                (c == ']' && top != '[') ||
                (c == '}' && top != '{')) return false;
        }
    }
    return st.empty();
}

int main() {
    std::stack<int> st;
    st.push(1);   // O(1)
    st.push(2);
    st.push(3);

    std::cout << "Top: " << st.top() << "\n";   // 3
    st.pop();                                    // O(1)
    std::cout << "After pop, top: " << st.top() << "\n"; // 2
    std::cout << "Size: " << st.size() << "\n";

    // Balanced parentheses demo
    std::cout << std::boolalpha;
    std::cout << "\"({[]})\" balanced: " << isBalanced("({[]})") << "\n";
    std::cout << "\"({[})\"  balanced: " << isBalanced("({[})")  << "\n";

    return 0;
}

File: computing/interview/data-structures/python/stack.py
# Stack — Python list (O(1) push/pop at end) or collections.deque
from collections import deque


def is_balanced(s: str) -> bool:
    """Classic use-case: balanced parentheses checker."""
    stack = []
    matching = {')': '(', ']': '[', '}': '{'}
    for ch in s:
        if ch in '([{':
            stack.append(ch)
        elif ch in ')]}':
            if not stack or stack[-1] != matching[ch]:
                return False
            stack.pop()
    return len(stack) == 0


# --- list as stack (append = push, pop = pop) ---
stack = []
stack.append(1)   # O(1) amortised
stack.append(2)
stack.append(3)
print("Top:", stack[-1])   # peek — O(1)
stack.pop()                # O(1)
print("After pop:", stack)

# --- deque as stack (slightly faster for large stacks) ---
dq = deque()
dq.append(10)
dq.append(20)
print("deque top:", dq[-1])
dq.pop()
print("deque after pop:", list(dq))

# --- Balanced parentheses demo ---
print(f'"({{[]}})" balanced: {is_balanced("({[]})")}')
print(f'"({{[})" balanced:  {is_balanced("({[})")}')
